### PR TITLE
add support for /dev/stdin and --output-dir to extract-paired-reads

### PR DESCRIPTION
**Pull request by ctb**
motivated by #847; addresses #1055 for extract-paired-reads.

### CODE REVIEW

**Comment by ctb** [comment]
- [ ] Is it mergeable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?
- [ ] Is the Copyright year up to date?

**Comment by ctb** [comment]
Ready for review & merge @mr-c @luizirber @camillescott

File: scripts/extract-paired-reads.py [hunk 1]
```diff
@@ -31,15 +31,29 @@
...
-    The output is two files, <input file>.pe and <input file>.se, placed in the
-    current directory. The .pe file contains interleaved and properly paired
-    sequences, while the .se file contains orphan sequences.
-
-    Many assemblers (e.g. Velvet) require that you give them either perfectly
-    interleaved files, or files containing only single reads. This script takes
-    files that were originally interleaved but where reads may have been
-    orphaned via error filtering, application of abundance filtering, digital
-    normalization in non-paired mode, or partitioning.
+    Separate properly interleaved reads from orphaned reads.
```
Comment (mr-c) [inline scripts/extract-paired-reads.py]: Any reason to repeat (though with different wording) the script summary at https://github.com/dib-lab/khmer/pull/1085/files#diff-7ba8eb069734923ae2a8b85fa4021e7eR63 here?

File: scripts/extract-paired-reads.py [hunk 2]
```diff
@@ -51,6 +65,19 @@ def get_parser():
     parser.add_argument('infile')
     parser.add_argument('--version', action='version', version='%(prog)s ' +
                         khmer.__version__)
+
+    parser.add_argument('-o', '--output-dir', metavar="output_directory",
+                        dest='output_directory', default='', help='Output '
```
Comment (mr-c) [inline scripts/extract-paired-reads.py]: Ideally the long form of an option would match the desired `args.name` which obviates the need to manually specify the `dest`. The `metavar` is used to override the `dest` name for aesthetic reasons. To my eyes the two lines above are rather redundant as ArgParse will convert dashes to underscores when generating the property name for this argument.

`parser.add_argument('-o', '--output-directory', default='',` and on as before should do the trick.

File: scripts/extract-paired-reads.py [hunk 3]
```diff
@@ -51,6 +65,19 @@ def get_parser():
...
+                        'split reads to specified directory. Creates '
+                        'directory if necessary')
+
+    parser.add_argument('-p', '--output-paired', metavar='output_paired',
+                        default=None, help='Output paired reads to this '
+                        'file', type=argparse.FileType('w'))
+    parser.add_argument('-s', '--output-single', metavar='output_single',
+                        default=None, help='Output orphaned reads to this '
```
Comment (mr-c) [inline scripts/extract-paired-reads.py]: Ditto my comments about metavar above.

File: scripts/extract-paired-reads.py (hunk 1 above)
Comment (ctb) [inline scripts/extract-paired-reads.py]: Removed.

File: scripts/extract-paired-reads.py [hunk 4]
```diff
@@ -60,25 +87,45 @@ def main():
...
-
-    print('reading file "%s"' % args.infile, file=sys.stderr)
-    print('outputting interleaved pairs to "%s.pe"' % outfile, file=sys.stderr)
-    print('outputting orphans to "%s.se"' % outfile, file=sys.stderr)
+    infile = args.infile
+    check_input_files(infile, args.force)
+    check_space([infile], args.force)
+
+    # decide where to put output files - specific directory? or just default?
+    if infile == '/dev/stdin':
```
Comment (mr-c) [inline scripts/extract-paired-reads.py]: Are we not accepting `-` for the infile name (or otherwise defaulting to stdin?)

**Comment by ctb** [comment]
Thx for the review; fixed @mr-c.

File: scripts/extract-paired-reads.py [hunk 5]
```diff
@@ -48,9 +60,22 @@ def get_parser():
     parser = argparse.ArgumentParser(
         description='Take a mixture of reads and split into pairs and '
         'orphans.', epilog=textwrap.dedent(epilog))
-    parser.add_argument('infile')
+    parser.add_argument('infile', nargs='?', default='/dev/stdin')
     parser.add_argument('--version', action='version', version='%(prog)s ' +
                         khmer.__version__)
+
+    parser.add_argument('-o', '--output-dir',
+                        dest='output_directory', default='', help='Output '
```
Comment (mr-c) [inline scripts/extract-paired-reads.py]: Thanks for removing the `metavars`. `dest` is still a bit redundant: one can either rename the option to `--output-directory` or change the references to be `args.output_dir` and the `dest` specification can be removed.

Comment (ctb) [inline scripts/extract-paired-reads.py]: On Sun, Jun 14, 2015 at 12:40:25PM -0700, Michael R. Crusoe wrote:
> >      parser.add_argument('--version', action='version', version='%(prog)s ' +
> >                          khmer.__version__)
> > +
> > +    parser.add_argument('-o', '--output-dir',
> > +                        dest='output_directory', default='', help='Output '
> 
> Thanks for removing the `metavars`. `dest` is still a bit redundant: one can either rename the option to `--output-directory` or change the references to be `args.output_dir` and the `dest` specification can be removed.

ok. I don't think this should hold up merges, frankly.

**Comment by ctb** [comment]
Updated @mr-c

**Comment by ctb** [comment]
whoops, fixed @mr-c.

**Comment by ctb** [comment]
Jenkins, please test this.

**Comment by ctb** [comment]
The failures seem to be on Jenkins side not on the code...?

**Comment by mr-c** [comment]
Jenkins, test this please

**Comment by mr-c** [comment]
LOL, I was wrong.

```
Identified problems

Matrix build khmer-pullrequest » linux #2357
ArgParse parameter of type FileType needs metavar workaround

Typically we fix this by adding `metavar="filename"` to the parameter definitions that use FileType
```

 I forgot that argparse.FileType arguments do need the metavar. Thank you Jenkins for reminding me :-P

**Comment by ctb** [comment]
Hah! Question: how do I trigger this error outside of Jenkins?

--
Titus Brown, ctbrown@ucdavis.edu

**Comment by mr-c** [comment]
`make doc`

**Comment by ctb** [comment]
No, I actually did that ;).  I get

     build succeeded, 2 warnings.

even when I blow away the build/sphinx/html directory. Any other thoughts?

**Comment by ctb** [comment]
Jenkins, please test this.

**Comment by ctb** [comment]
Jenkins, please test this

**Comment by ctb** [comment]
Jenkins, test this please

**Comment by ctb** [comment]
test this please, jenkins

**Comment by ctb** [comment]
@mr-c, now getting this error:

````
 test_save_load_large

Traceback (most recent call last): File "/usr/lib/python2.7/unittest/case.py", line 331, in run testMethod() File "/var/lib/jenkins/workspace/khmer-pullrequest@2/label/linux/.env/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest self.test(*self.arg) File "/var/lib/jenkins/workspace/khmer-pullrequest@2/label/linux/tests/test_counting_hash.py", line 591, in test_save_load_large do_test(ctfile) File "/var/lib/jenkins/workspace/khmer-pullrequest@2/label/linux/tests/test_counting_hash.py", line 579, in do_test orig = khmer.CountingHash(12, sizes) MemoryError 
```

**Comment by mr-c** [comment]
@ctb there may be a memory leak in `test_counting_hash.test_get_kmers` I'm investigating it now.

**Comment by mr-c** [comment]
Jenkins, test this please

**Comment by kdmurray91** [comment]
@ctb that error is due to a test which makes a Countgraph w/ 1 table of 2e9 and a bit, to test the issue with an infinite loop in gzip, from  #1043. I can remove that test if need be.

**Comment by ctb** [comment]
@kdmurray91 I know where the test comes from - I think the question is, should this be run by Jenkins regularly or not? We now have a few "big" tests that we might want to run only when getting ready for a release.

**Comment by ctb** [comment]
Ready for review & merge.

File: ChangeLog [hunk 6]
```diff
@@ -13,6 +13,16 @@
 
 2015-06-14  Titus Brown  <titus@idyll.org>
 
+   * scripts/extract-paired-reads.py: added --output_dir, --paired-output,
```
Comment (mr-c) [inline ChangeLog]: Should be top of ChangeLog (can be fixed during actual merge).

Comment (ctb) [inline ChangeLog]: ok shall I make this mergable/fix ChangeLog, or are there other things that
need to be done?

Comment (mr-c) [inline ChangeLog]: I tried doing a coverage check on my laptop but it said all changed  lines
were uncovered which seems unlikely.

On Fri, Jun 26, 2015, 17:30 C. Titus Brown <notifications@github.com> wrote:

> In ChangeLog
> <https://github.com/dib-lab/khmer/pull/1085#discussion_r33406746>:
>
> > @@ -13,6 +13,16 @@
> >
> >  2015-06-14  Titus Brown  <titus@idyll.org>
> >
> > +   * scripts/extract-paired-reads.py: added --output_dir, --paired-output,
>
> ok shall I make this mergable/fix ChangeLog, or are there other things
> that need to be done?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dib-lab/khmer/pull/1085/files#r33406746>.
>
-- 
Michael R. Crusoe: Programmer & Bioinformatician crusoe@ucdavis.edu
The lab for Data Intensive Biology; University of California, Davis
https://impactstory.org/MichaelRCrusoe http://twitter.com/biocrusoe

**Comment by mr-c** [comment]
LGTM!

**Comment by ctb** [comment]
woot, tnx!

**Pull request merged**